File: rush-todo-web/src/validators/WorkItemValidator.ts
import type { WorkItemModel } from '../api/models'
import { DateValidator, TextValidator, ValidatorBase } from './ValidatorBase'

class WorkItemValidator extends ValidatorBase {
  readonly address: TextValidator
  readonly description: TextValidator
  readonly scheduledDate: DateValidator
  readonly title: TextValidator

  constructor(workItem: WorkItemModel) {
    super('Work item')

    this.title = new TextValidator({
      getValue: () => workItem.title,
      label: 'Title',
      maxLength: 255,
      validation: (value) => validateRequiredText(value, 255),
    })

    this.description = new TextValidator({
      getValue: () => workItem.description,
      label: 'Description',
    })

    this.address = new TextValidator({
      getValue: () => workItem.address,
      label: 'Address',
      maxLength: 255,
      validation: (value) => validateRequiredText(value, 255),
    })

    this.scheduledDate = new DateValidator({
      getValue: () => workItem.scheduledDate,
      label: 'Scheduled date',
    })
  }

  override validate(): string | undefined {
    return (
      this.title.validate() ?? this.address.validate()
    )
  }
}

function validateRequiredText(value: string | null, maxLength: number): string | undefined {
  if (!value?.trim()) return '[Label] is required'
  return value.length > maxLength
    ? `[Label] cannot exceed ${maxLength} characters`
    : undefined
}

export default WorkItemValidator

File: RushTodo.Api/Validators/WorkItemValidator.cs
using RushTodo.Api.Entities;
using RushTodo.Api.Models;

namespace RushTodo.Api.Validators;

public class WorkItemValidator(WorkItemModel workItem) : ValidatorBase<WorkItemModel>(workItem)
{
    public TextValidator Title => new()
    {
        Label = "Title",
        GetValue = () => Model.Title,
        MaxLength = 255,
        Validation = value => ValidateRequiredText(value, 255),
    };

    public TextValidator Description => new()
    {
        Label = "Description",
        GetValue = () => Model.Description,
    };

    public TextValidator Address => new()
    {
        Label = "Address",
        GetValue = () => Model.Address,
        MaxLength = 255,
        Validation = value => ValidateRequiredText(value, 255),
    };

    public DateValidator ScheduledDate => new()
    {
        Label = "Scheduled date",
        GetValue = () => Model.ScheduledDate,
        IsVisible = Model.StatusId == WorkItemStatusId.Scheduled,
        Validation = value => value is null ? "[Label] is required when status is Scheduled" : null,
    };

    public DateValidator CompletionDate => new()
    {
        Label = "Completion date",
        GetValue = () => Model.CompletionDate,
        IsVisible = Model.StatusId == WorkItemStatusId.Done,
        Validation = value => value is null ? "[Label] is required when status is Done" : null,
    };

    public DateValidator CancellationDate => new()
    {
        Label = "Cancellation date",
        GetValue = () => Model.CancellationDate,
        IsVisible = Model.StatusId == WorkItemStatusId.Cancelled,
        Validation = value => value is null ? "[Label] is required when status is Cancelled" : null,
    };

    public override string? Validate()
        => Title.Validate()
           ?? Address.Validate()
           ?? ScheduledDate.Validate()
           ?? CompletionDate.Validate()
           ?? CancellationDate.Validate();

    private static string? ValidateRequiredText(string? value, int maxLength)
    {
        if (string.IsNullOrWhiteSpace(value)) return "[Label] is required";
        return value.Length > maxLength ? $"[Label] cannot exceed {maxLength} characters" : null;
    }
}
